graphql_compiler/tests/integration_tests/test_backends_integration.py

```diff
@@ -555,6 +555,129 @@ def test_recurse(self) -> None:
+                    Animal {
+                        name @output(out_name: "animal")
+                        in_Animal_ParentOf {
+                            name @output(out_name: "father")
+                        }
+                    }
+                }""",
+                [
+                    {"animal": "Animal 1", "father": "Animal 1"},
+                    {"animal": "Animal 2", "father": "Animal 1"},
+                    {"animal": "Animal 3", "father": "Animal 1"},
+                    {"animal": "Animal 4", "father": "Animal 3"},
+                ],
+            ),
+            # Query 2: Get all grandparents
+            (
+                """
+                {
+                    Animal {
+                        name @output(out_name: "animal")
+                        in_Animal_ParentOf {
+                            in_Animal_ParentOf {
+                                name @output(out_name: "grandfather")
+                            }
+                        }
+                    }
+                }""",
+                [
+                    {"animal": "Animal 1", "grandfather": "Animal 1"},
+                    {"animal": "Animal 2", "grandfather": "Animal 1"},
+                    {"animal": "Animal 3", "grandfather": "Animal 1"},
+                    {"animal": "Animal 4", "grandfather": "Animal 1"},
+                ],
+            ),
+            # Query 3: Use recursion to get self, parent and grandparent
+            (
+                """
+                {
+                    Animal {
+                        name @output(out_name: "animal")
+                        in_Animal_ParentOf @recurse(depth: 2) {
+                            name @output(out_name: "ancestor")
+                        }
+                    }
+                }""",
+                [
+                    {"animal": "Animal 1", "ancestor": "Animal 1"},  # self
+                    {"animal": "Animal 1", "ancestor": "Animal 1"},  # parent
+                    {"animal": "Animal 1", "ancestor": "Animal 1"},  # grandparent
+                    {"animal": "Animal 2", "ancestor": "Animal 2"},  # self
+                    {"animal": "Animal 2", "ancestor": "Animal 1"},  # parent
+                    {"animal": "Animal 2", "ancestor": "Animal 1"},  # grandparent
+                    {"animal": "Animal 3", "ancestor": "Animal 3"},  # self
+                    {"animal": "Animal 3", "ancestor": "Animal 1"},  # parent
+                    {"animal": "Animal 3", "ancestor": "Animal 1"},  # grandparent
+                    {"animal": "Animal 4", "ancestor": "Animal 4"},  # self
+                    {"animal": "Animal 4", "ancestor": "Animal 3"},  # parent
+                    {"animal": "Animal 4", "ancestor": "Animal 1"},  # grandparent
+                ],
+            ),
+            # Query 4: Unfold recursion to omit self
+            #
+            #  XXX: This output is incorrect. I'm just showing the current output. Results
```

Comment:
This might be better split up off into its own test suite, with each test query as a separate test function, and this last one as an `xfail` test (with the correct input) until we can fix it.

Reply:
The point of the first 3 tests is to run before the fourth. I'm testing recurse results in relation to non-recurse results:
- If the data changes and test 4 fails, likely test 3 will fail as well, pointing to the issue instead of suggesting recurse might be broken.
- If the data changes and test 4 passes, but is no longer a good test for duplication, test 3 will likely fail.